iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/flush/Scheduler.java

```diff
@@ -0,0 +1,261 @@
+        .join();
+  }
+
+  /**
+   * Keep fetching evictable nodes from cacheManager until the memory status is under safe mode or
+   * no node could be evicted. Update the memory status after evicting each node.
+   *
+   * @param force true if force to evict all cache
+   */
+  public synchronized void scheduleRelease(boolean force) {
+    CompletableFuture.allOf(
+            regionToStore.values().stream()
+                .map(
+                    store ->
+                        CompletableFuture.runAsync(
+                            () -> {
+                              while (force || releaseFlushStrategy.isExceedReleaseThreshold()) {
+                                // store try to release memory if not exceed release threshold
+                                if (store.executeMemoryRelease()) {
+                                  // if store can not release memory, break
+                                  break;
+                                }
+                              }
+                            },
+                            workerPool))
+                .toArray(CompletableFuture[]::new))
+        .join();
+  }
+
+  /**
+   * Select some subtrees to flush. The subtrees are selected from the MTreeStore by the sequence of
+   * the regionIds. The number of subtrees to flush is determined by parameter {@link
+   * Scheduler#BATCH_FLUSH_SUBTREE}. It will return asynchronously. If worker pool is full, the task
+   * will be discarded directly.
+   *
+   * @param regionIds determine the MTreeStore to select subtrees, the head of the list is the first
+   *     MTreeStore to select subtrees
+   */
+  public synchronized void scheduleFlush(List<Integer> regionIds) {
+    AtomicInteger remainToFlush = new AtomicInteger(BATCH_FLUSH_SUBTREE);
+    for (int regionId : regionIds) {
+      if (flushingRegionSet.contains(regionId)) {
+        continue;
+      }
+      flushingRegionSet.add(regionId);
+      workerPool.submit(
+          () -> {
+            CachedMTreeStore store = regionToStore.get(regionId);
+            ICacheManager cacheManager = store.getCacheManager();
+            ISchemaFile file = store.getSchemaFile();
+            LockManager lockManager = store.getLockManager();
+            List<ICachedMNode> nodesToFlush = new ArrayList<>();
+            PBTreeFlushExecutor flushExecutor;
+            long startTime = System.currentTimeMillis();
+            try {
+              IDatabaseMNode<ICachedMNode> dbNode = cacheManager.collectUpdatedStorageGroupMNodes();
+              if (dbNode != null) {
+                flushExecutor = new PBTreeFlushExecutor(dbNode, cacheManager, file, lockManager);
+                flushExecutor.flushDatabase();
+                remainToFlush.decrementAndGet();
+              }
+              Iterator<ICachedMNode> volatileSubtrees = cacheManager.collectVolatileSubtrees();
+              while (volatileSubtrees.hasNext()) {
+                nodesToFlush.add(volatileSubtrees.next());
```

Comment:
Need more discussion here. 

Such behavior may lock all the collected node, and none R/W operation will be allowed to be executed under the related subtree until the subtree has been flushed.

Reply:
It seems that encapsulating the logic of taking the node out of the iterator in the PBTreeFlushExecutor would solve the problem of holding locks. 

But there will be a new problem: Schedule will have a counter for restoreToFlush to record the subtreeRoots of each timed task flush. Ff the logic is encapsulated, there is no way for the Scheduler to sense it in advance. Can the CacheManager give an interface to estimate the number of subtreeRoots in an iterator?